### PR TITLE
Flow

### PR DESCRIPTION
**Pull request by seelabs**
This is a new implementation of payments. 

The key changes from the old code are:

* A new algorithm for finding the liquidity in a payment path. There is still a reverse and forward pass, but the forward pass starts at the limiting step rather than the payment source. This insures the limiting step is completely consumed rather than potentially leaving a 'dust' amount in the forward pass.

* Each step in a payment is either a book step, a direct step (account to account step), or an xrp endpoint. Each step in the existing implementation is a triple, where each element in the triple is either an account of a book, for a total of eight step types.

* Because we consider accounts in pairs, rather than triples, transfer fees are handled differently. In the existing code, in the payment path A -> gw ->B, if A redeems to gw, and gw issues to B, a transfer fee is changed. In the new code, a transfer fee is changed even if A issues to gw.

The commits are organized to be reviewed in order.

@nbougalis @scottschurr

### CODE REVIEW

**Comment by scottschurr** [comment]
This was a massive undertaking.  I wish I understood the corners of payments well enough to give this a truly thorough review.  But I did the best I could.  Here are three things we should talk about before I give a thumbs up.

1. There are a couple of places in inheritance trees that I suspect need a `virtual` destructor.

2. What I really wished for during the review was documentation (in English) of all of the corners of your payments implementation.  I'd wish for that to include motivations for all of the various sub-classes of `Step` and descriptions of how a Flow and a Strand are intended to work.  By the time I got to the end of the review I had a vague idea.  But I wish I'd been more confident going in.  This documentation would be useful in the future as we do maintenance and more people need to understand this code.

3. I'm concerned about the possible performance issues with the `dynamic_cast`s.  Consider putting something in the virtual interface of `Step` that can make those `dynamic_cast`s go away.  I have some ideas but I suspect that if you put your mind to it your ideas would be better.

Nice work.  Yes, it's complicated.  But it's a complicated problem.  If I had suggestions for how to simplify it I'd offer those.  But I don't.

**Comment by seelabs** [comment]
I pushed a set of commits that addresses almost all the PR feedback (the commits from `Address feedback from PR` to `Remove or replace log param` are new). I still want to carefully review the `mulRatio` functions, and I still need to address what to do about Nik's `Simplify calculation of transfer fee` PR.

File: src/ripple/app/paths/impl/StepChecks.h [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        // check freeze
+        if (auto sle = view.read (keylet::account (dst)))
+        {
+            if (sle->isFlag (lsfGlobalFreeze))
+            {
+                return terNO_LINE;
+            }
+        }
+
```
Comment (JoelKatz) [inline src/ripple/app/paths/impl/StepChecks.h]: I'm not sure what the case is where `src == dst`. Maybe book->account or account->book? If that's not possible/valid, we should probably return an error. If it is possible, I'm not sure it's okay to skip the check for global freeze. (Obviously, we can skip the check for a frozen line, you can't have a line to yourself.)

**Comment by scottschurr** [comment]
The commits that address feedback look good to me, although I did leave a few notes around.  I know you're still intending to spend some quality time with `mulRatio()`.  And I'm hoping that some sort of overview documentation is still coming.  But other than that it looks good.

File: src/ripple/app/paths/impl/StepChecks.h (hunk 1 above)
Comment (seelabs) [inline src/ripple/app/paths/impl/StepChecks.h]: `src == dst` should not happen. Book steps do not use this check (book steps see frozen offer owners as having a zero balance). It doesn't make sense to have a direct step with the same src and dst. I changed the `if` statement to an assert.

**Comment by seelabs** [comment]
Rebased onto b13. Also addressed @scottschurr typo and cleaned up `IOUAmount::mulRatio`.

File: src/ripple/app/paths/impl/XRPEndpointStep.cpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    {
+        if (!cache_)
+            return boost::none;
+        return EitherAmount (*cache_);
+    }
+
+    boost::optional<EitherAmount>
+    cachedOut () const override
+    {
+        if (!cache_)
```
Comment (seelabs) [inline src/ripple/app/paths/impl/XRPEndpointStep.cpp]: Added a `cached()` helper function that both `cachedIn` and `cachedOut` call and added a comment to `cache~` as to why we use one cache value for both in and out.

File: src/ripple/app/paths/impl/PaySteps.cpp [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+            JLOG (j.warning) << "toStep failed";
+            return {s.first, Strand{}};
+        }
+    }
+
+    return {tesSUCCESS, std::move (result)};
+}
+
+std::pair<TER, std::vector<Strand>>
+toStrands (
```
Comment (nbougalis) [inline src/ripple/app/paths/impl/PaySteps.cpp]: So many lambdas (and lambdas embedded in lambdas) between this function and the one immediately before it that it's hard to follow. I'm wondering if we could get slightly more readable code by abstracting those two functions to a class, with private members and a single public interface.

Comment (seelabs) [inline src/ripple/app/paths/impl/PaySteps.cpp]: Most often, I use lambdas for two things:

1) Complex initialization of a variable. This let's us replace code like:
```
Foo var;
{
   // Some code
   var = blockResult;
}
```
with:
```
auto const var = [&]
{
    // Some code
    return fooBlockResult;
}();
```

2) Local functions that require use of lots of the local function state. For example:
```
Foo(...a whole bunch of parameters...)
{
    // ...a whole bunch of local state
    auto helperFun = [&] {
        // ...use the parameters and local state
    }

    if (something)
    {
        helperFun();
    }

    while(something)
    {
        helperFun();
    }
}
```
If I made `helperFun` a free function I would have to pass a large number of parameters to it. The other alternative is a class to wrap local state and make `helperFun` a member var on that class.

Of these two cases, I think (1) is almost always OK. It's very easy to see what's going on. (2) is more of a problem. I should only use (2) when there are no other alternatives.

If I can refactor the code so the lambda doesn't need to capture so much local state, I can replace the lambda with a free function or small class.

Enough generalities, let's talk about this specific case.

What I'm doing with the `add` lambda is adding a strand to a `vector`, but I'm using the `vector` like a `set`. I create the strand, check if it's not already part of the set, and add it if it's not. I see three ways to improve this code:

1) Move the call `toStrand` outside the lambda. This means I no longer have to capture as many locals.

2) Write an `bool operator==(Strand const& lhs, Strand const& rhs)` free function. This removes one of the local lambdas and lets me use the `find` algorithm, which everyone knows.

3) Create a `StrandSet` class. This makes it clear what `add` is meant to do.

Note that I can't use something like `boost::flat_set` because I don't have `operator<` defined on `Step`. I did have an `operator<` defined at one point, but my implementation used `info` classes which was somewhat complex.

I also understand your comment wasn't necessarily about just this lambda. I'll be more thoughtful when using lambdas - especially type (2) lambdas.

File: src/ripple/app/paths/impl/PaySteps.cpp [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+    STPathElement const firstNode (
+        STPathElement::typeAll, src, curIssue.currency, curIssue.account);
+
+    boost::optional<STPathElement> sendMaxPE;
+    if (sendMaxIssue && sendMaxIssue->account != src)
+        sendMaxPE.emplace (sendMaxIssue->account, boost::none, boost::none);
+
+    STPathElement const lastNode (dst, boost::none, boost::none);
+
+    auto isAccount = [](STPathElement const* pe)
```
Comment (vinniefalco) [inline src/ripple/app/paths/impl/PaySteps.cpp]: These five lambdas could be member functions on `STPathElement`. Although I don't have strong feelings either way.

File: src/ripple/app/paths/impl/PaySteps.cpp [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+        pes.push_back (&i);
+    auto const lastCurrency=[&]{
+        for(auto i=pes.crbegin(), e=pes.crend();i!=e;++i)
+        {
+            if (hasCurrency (*i))
+                return (*i)->getCurrency();
+        }
+        assert(0);
+        return deliver.currency;
+    }();
```
Comment (vinniefalco) [inline src/ripple/app/paths/impl/PaySteps.cpp]: Consider rewriting 213-221 as an invariant check followed by a standard algorithm:
```
assert(boost::find_if(pes, [](auto const& pe) { return pe->hasCurrency(); }) != pes.end());
```

File: src/ripple/app/paths/impl/PaySteps.cpp [hunk 6]
```diff
@@ -0,0 +1,533 @@
...
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+            }
+            else if (isXRP (nextCurrency))
+            {
+                JLOG (j.trace) << "Inserting implied IX offer";
+                auto msr = make_BookStepIX (ctx(), curIssue);
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
```
Comment (vinniefalco) [inline src/ripple/app/paths/impl/PaySteps.cpp]: Is there a difference from
`result.emplace_back(std::move (msr.second));` ?
Which should we use?

Comment (HowardHinnant) [inline src/ripple/app/paths/impl/PaySteps.cpp]: I did a study on this and there is no difference between `push_back` and `emplace_back` when the single argument is the same type as the `vector::value_type`:  http://htmlpreview.github.io/?https://github.com/HowardHinnant/papers/blob/master/insert_vs_emplace.html  `insert` vs `emplace` is far more complicated though.

File: src/ripple/app/paths/impl/PaySteps.cpp (hunk 3 above)
Comment (seelabs) [inline src/ripple/app/paths/impl/PaySteps.cpp]: Refactored to remove some lambdas

File: src/ripple/app/paths/impl/PaySteps.cpp (hunk 5 above)
Comment (seelabs) [inline src/ripple/app/paths/impl/PaySteps.cpp]: Done. And we don't need the `assert`, I removed it.

File: src/ripple/app/paths/impl/PaySteps.cpp (hunk 4 above)
Comment (seelabs) [inline src/ripple/app/paths/impl/PaySteps.cpp]: I got rid of all of these except `hasCurrency`, which I only used as an argument to `find_if`. I replaced the other uses with a member function on `STPathElement`.

**Comment by seelabs** [comment]
@nbougalis @scottschurr I pushed two new commits that address the latest round of comments. I also rebased to b14.

**Comment by scottschurr** [comment]
The CI failures look legit.  I suspect a missing include in the nounity builds.

**Comment by seelabs** [comment]
@scottschurr Thanks for alerting me to the nonunity build failure. I push a commit that fixes that and also responds to the feedback so far in the previous commits.

**Comment by nbougalis** [comment]
@seelabs don't forget to consider what you want to do with #1533

**Comment by seelabs** [comment]
@nbougalis I incorporated Nik's PR from https://github.com/ripple/rippled/pull/1533 into this PR. I made some changes (I added a switchover and kept the old behavior until we switchover to the new flow code). That last commit needs to be reviewed @nbougalis @scottschurr

**Comment by seelabs** [comment]
@nbougalis @scottschurr I pushed two commits today: one to address the previous feedback and add Nik's fess PR to this PR and one to limit the number of offers a step will consume.

The second commit also removes an assert that could falsely trigger in some cases. I upped the logging to `error` in that case, though it will only do the check in debug mode.

**Comment by nbougalis** [comment]
Looks good to me. :+1:

File: src/ripple/app/paths/impl/Steps.h [hunk 7]
```diff
@@ -37,11 +37,26 @@ class ApplyView;
...
+     BookStepXI is an XRP/IOU offer book
+     XRPEndpointStep is the source or destination account for XRP
+
+   Amounts may be transformed through a step in either the forward or the
+   reverse direction. In the forward direction, the funcion `fwd` is used to
+   find the amount the step would output given an input amount. In the reverse
+   direction, the function `rev` is used to find the amount of input needed to
+   produce the desired output.
+
+   Amounts are always transfromed using liquidity with the same quality (quality
```
Comment (scottschurr) [inline src/ripple/app/paths/impl/Steps.h]: transfromed -> transformed?

Comment (vinniefalco) [inline src/ripple/app/paths/impl/Steps.h]: transfromed = transfered from
get with dar lingo YO!

File: src/ripple/protocol/tests/XRPAmount.test.cpp [hunk 8]
```diff
@@ -126,19 +126,30 @@ class XRPAmount_test : public beast::unit_test::suite
         }
 
         {
+            // Similar test as above, but for neative values
```
Comment (scottschurr) [inline src/ripple/protocol/tests/XRPAmount.test.cpp]: neative -> negative?

Comment (vinniefalco) [inline src/ripple/protocol/tests/XRPAmount.test.cpp]: A portmanteau of "neat" and "live?"

Comment (scottschurr) [inline src/ripple/protocol/tests/XRPAmount.test.cpp]: "portmanteau" is an excellent word.  I'll need to file that one away.

**Comment by scottschurr** [comment]
:+1: Looks good to me.  Thanks for the additional comments.

File: src/ripple/app/paths/impl/Steps.h (hunk 7 above)
Comment (seelabs) [inline src/ripple/app/paths/impl/Steps.h]: Fixed

**Comment by nbougalis** [comment]
This has passed review. @seelabs, can you squash and rebase on the latest?

**Comment by seelabs** [comment]
@nbougalis Will do

**Comment by seelabs** [comment]
Rebased and squashed
